File: runtime/typescript/packages/core/src/core/interfaces.ts
/**
 * Plugin interfaces for the Prompty pipeline.
 *
 * Each pipeline step is defined as an interface. Implementations are
 * registered via the registry and discovered at runtime by key
 * (e.g., "nunjucks" for renderers, "openai" for executors).
 *
 * @module
 */

import type { Prompty } from "../model/prompty.js";
import type { Message } from "./types.js";

// ---------------------------------------------------------------------------
// Renderer
// ---------------------------------------------------------------------------

/**
 * Renders a template string with the given inputs.
 *
 * Discovered by: `agent.template.format.kind` (e.g., "nunjucks", "mustache").
 */
export interface Renderer {
  render(agent: Prompty, template: string, inputs: Record<string, unknown>): Promise<string>;
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

/**
 * Parses a rendered string into an array of abstract Messages.
 *
 * Discovered by: `agent.template.parser.kind` (e.g., "prompty").
 *
 * Optionally implements `preRender()` for nonce injection (strict mode).
 */
export interface Parser {
  parse(agent: Prompty, rendered: string, context?: Record<string, unknown>): Promise<Message[]>;

  /**
   * Optional hook called before rendering to sanitize the template
   * and produce context used during parsing.
   *
   * Returns `[sanitizedTemplate, context]`.
   */
  preRender?(template: string): [string, Record<string, unknown>];
}

// ---------------------------------------------------------------------------
// Executor
// ---------------------------------------------------------------------------

/**
 * Sends messages to an LLM provider and returns the raw response.
 *
 * Discovered by: `agent.model.provider` (e.g., "openai", "azure").
 */
export interface Executor {
  execute(agent: Prompty, messages: Message[]): Promise<unknown>;
}

// ---------------------------------------------------------------------------
// Processor
// ---------------------------------------------------------------------------

/**
 * Extracts clean results from raw LLM responses.
 *
 * Discovered by: `agent.model.provider` (e.g., "openai", "azure").
 */
export interface Processor {
  process(agent: Prompty, response: unknown): Promise<unknown>;
}

File: runtime/csharp/Prompty.Core/Interfaces.cs
// Copyright (c) Microsoft. All rights reserved.

namespace Prompty.Core;

/// <summary>
/// Renders a PromptAgent's instructions template with input values.
/// Registered by template format kind (e.g., "jinja2", "mustache").
/// </summary>
public interface IRenderer
{
    Task<string> RenderAsync(Prompty agent, string template, Dictionary<string, object?> inputs);
}

/// <summary>
/// Parses rendered text into a list of Messages.
/// Registered by parser kind (e.g., "prompty").
/// </summary>
public interface IParser
{
    Task<List<Message>> ParseAsync(Prompty agent, string rendered);
}

/// <summary>
/// Optional interface for parsers that can sanitize templates before rendering.
/// When implemented, the pipeline calls PreRender first to get a cleaned template
/// and context dict, then renders, then parses with that context.
/// </summary>
public interface IPreRenderable
{
    (string template, Dictionary<string, object?> context) PreRender(string template);
}

/// <summary>
/// Executes an LLM call with prepared messages.
/// Registered by provider name (e.g., "openai", "foundry", "anthropic").
/// </summary>
public interface IExecutor
{
    Task<object> ExecuteAsync(Prompty agent, List<Message> messages);

    /// <summary>
    /// Formats the assistant tool-call response and dispatched tool results into
    /// messages for the next agent loop iteration. Each provider implements this
    /// to match its API's expected wire format, keeping the pipeline provider-agnostic.
    /// </summary>
    /// <param name="rawResponse">Original LLM response (for content block preservation).</param>
    /// <param name="toolCalls">Tool calls extracted by the processor.</param>
    /// <param name="toolResults">Results from dispatching each tool call, parallel to toolCalls.</param>
    /// <param name="textContent">Any non-tool text content from the response.</param>
    List<Message> FormatToolMessages(
        object rawResponse,
        List<ToolCall> toolCalls,
        List<ToolResult> toolResults,
        string? textContent = null);
}

/// <summary>
/// Post-processes raw LLM responses into a final result.
/// Registered by provider name (e.g., "openai", "foundry", "anthropic").
/// </summary>
public interface IProcessor
{
    Task<object> ProcessAsync(Prompty agent, object response);
}
